### ISSUE TITLE
minor error in the text about profiles

### ISSUE THREAD

**Issue by lumpsoid** [ibhagwan/fzf-lua#664]
Was playing with the plugin, and because this error occurs the first time profiles are mentioned, it led to a bit of confusion.

There is a minor error in the text about Profiles.
- [on the main page](https://github.com/ibhagwan/fzf-lua#profiles)
- [and in the lua profiles](https://github.com/ibhagwan/fzf-lua/tree/main/lua/fzf-lua/profiles#fzflua-profiles)

```
diff --git a/README.md b/README.md
index 3687402..595092b 100644
--- a/README.md
+++ b/README.md
@@ -342,7 +342,7 @@ Conviniently, fzf-lua comes with a set of preconfigured profiles, notably:
 Use `:FzfLua profiles` to experiment with the different profiles, once you've found what
 you like and wish to make the profile persist, send a `string` argument to `setup`:

-require('fzf-lua').setup('fzf-native')
+require('fzf-lua').setup({'fzf-native'})

 > **Note:** `setup` can be called multiple times for profile "live" switching

diff --git a/lua/fzf-lua/profiles/README.md b/lua/fzf-lua/profiles/README.md
index 5424459..910de55 100644
--- a/lua/fzf-lua/profiles/README.md
+++ b/lua/fzf-lua/profiles/README.md
@@ -4,7 +4,7 @@ This folder contains preconfigured fzf-lua profiles which can be activated using
 `:FzfLua profiles` or used as a string argument directly in the `setup` function,
 for example:

-require("fzf-lua").setup("fzf-native")
+require("fzf-lua").setup({"fzf-native"})

 > **Note:** `setup` can be called multiple times for profile "live" switching
```

**Comment by ibhagwan** [comment]
That’s not an error, when calling setup with different options you should be sending a table, I.e. `setup({ … })`, to use a preconfigured profile you send a string argument, I.e. `setup('fzf-lua')`, no other options can be sent with it, must be a string for the profile code to be loaded.

**Comment by lumpsoid** [comment]
trying run this code
```
:lua require('fzf-lua').setup('fzf-native')
```
getting
![image](https://user-images.githubusercontent.com/79290581/219961243-b338f28c-3841-4f6f-9a07-4c8f0773ce38.png)


```
:lua require('fzf-lua').setup({'fzf-native'})
```
![image](https://user-images.githubusercontent.com/79290581/219961284-ef5c458e-8195-4ad2-a75b-f4a089d19bbd.png)
and this variant is running just fine.

**Comment by ibhagwan** [comment]
I think the error is coming from somewhere else, **I believe your fzf-lua isn't updated to the latest commit** but it might not beed a bad idea to change the profiles to work as you suggested (first argument **in a table**) so it's less error prone.

What's your current fzf-lua commit?

> and this variant is running just fine

Yes, but that one is doing nothing ATM :-)

**Comment by lumpsoid** [comment]
> That’s not an error, when calling setup with different options you should be sending a table, I.e. setup({ … }), to use a preconfigured profile you send a string argument, I.e. setup('fzf-lua')

oh,  I'm sorry, now I see the difference. In other places in the documentation, setup refers to a table because it directly defines the parameter we want to change.

but my understanding was confused because I couldn't use the pre-created profiles through the string.

> I believe your fzf-lua isn't updated to the latest commit

and I installed it yesterday, or are you talking about last last commit?

> Yes, but that one is doing nothing ATM :-)

yea, this  was my concearn )))))

> it's less error prone

in terms of whether any error would have been raised in this way, rather than me just being let through without any notification?

**Comment by ibhagwan** [comment]
> in terms of whether any error would have been raised in this way, rather than me just being let through without any notification?

The errors/notifications are purposfully supressed when called from `setup` as I believed people calling a profile from setup would wind up being annoyed with the messages every time they start neovim, the messages do appear when running `:FzfLua profiles` so you can see if the profile works as expected and you like it.

> and I installed it yesterday, or are you talking about last last commit?

That should work then, let me commit the changes your purposed and we'll try again :-)

**Comment by lumpsoid** [comment]
when I running `:FzfLua profiles`
```
E5108: Error executing lua .../nvim/site/pack/packer/start/fzf-lua/lua/fzf-lua/cmd.lua:78: vim/shared.lua:0
: after the second argument: expected table, got nil
stack traceback:
        [builtin#36]: at 0x7f7a4dee38c0
        .../nvim/site/pack/packer/start/fzf-lua/lua/fzf-lua/cmd.lua:78: in function 'run_command'
        .../nvim/site/pack/packer/start/fzf-lua/lua/fzf-lua/cmd.lua:100: in function 'load_command'
        [string ":lua"]:1: in main chunk
Press ENTER or type command to continue
```

**Comment by ibhagwan** [comment]
Seems like a bug, not sure where it's coming from still, does everything else work (files, grep, etc)?
What does `:FzfLua builtin` return? Do you see `profiles` listed?

**Comment by lumpsoid** [comment]
I apologise in advance for the off topic.

but I'm curious about design of how you return selected item after user press <enter> on th item he wants to select.

in telescope it will return table with each data in separate `key: value` pair, and here you just go all in, and eventualy you parse this hole string for the needed parts in order to navigate to this file.

**Comment by lumpsoid** [comment]
> What does :FzfLua builtin return? 

Hang on a second, something's been messed up, I'll find out.

**Comment by lumpsoid** [comment]
Ahem, yeah.
something was broken... something was fixed by restart.
yeah

Now everything is working perfectly

**Comment by ibhagwan** [comment]
> Now everything is working perfectly

Can you now switch profiles properly? both using `:FzfLua profiles` and `setup`?
The latest commits, with your suggestion makes it so you can use a profile as "baseline" and add your options on top, for example, telescope defaults with `bat` previewer:
```lua
:lua require"fzf-lua".setup({"telescope",winopts={preview={default="bat"}}})
```

> but I'm curious about design of how you return selected item after user press on th item he wants to select.

As for this, if you meant what happens behind the scenes: `fzf` prints the results to stdout, fzf-lua parses the results into a lua table which you can act upon using the `actions` table.

**Comment by lumpsoid** [comment]
> The [advacned Wiki](https://github.com/ibhagwan/fzf-lua/wiki/Advanced#fzf-exec-acts) is your friend

I mean why did you design it that way, because when you click on the selected element `selected[1]` equals the string we see in the search. Why not take the design of the telescope plugin, for example, where a table is returned where each part of the return item is in its own `key: value` pair.

because in the end the function parses this string anyway to get all the values needed to open the file to which the selected answer belongs.

**Comment by ibhagwan** [comment]
> > The [advacned Wiki](https://github.com/ibhagwan/fzf-lua/wiki/Advanced#fzf-exec-acts) is your friend
> 
> I mean why did you design it that way, because when you click on the selected element `selected[1]` equals the string we see in the search. Why not take the design of the telescope plugin, for example, where a table is returned where each part of the return item is in its own `key: value` pair.
> 
> because in the end the function parses this string anyway to get all the values needed to open the file to which the selected answer belongs.

What if you selected multiple items? A map does not maintain order, you'd need to iterate the map with `pairs` (as opposed to `ipairs`) and lose the ordering, this way seems most intuitive and straight forward that works for all use case (not just files, grep results), it gives you back **exactly** what fzf outputs which is easy to understand if you've been using fzf in the shell as well.

TL;DR, I tried to keep it as close as possible to fzf in the shell and `fzf.vim` APIs.

**Comment by lumpsoid** [comment]
indeed, there are also multiple choice options, and for consistency yes, this makes sense
Yes, thanks for the explanation

> As for this, if you meant what happens behind the scenes

I was digging around in the plugin's code to figure out what was going on, to customise the plugin, so, as I had to parse the string by hand, I remembered how telescope did it, when I also dug around in their code, I was surprised that they organised it through tables there

> Can you now switch profiles properly? both using :FzfLua profiles and setup?

is there a 100% way of knowing which profile is being used? because I not sure about `:lua require"fzf-lua".setup({"fzf-native")` version, it seems it still don't work properly

as for `:lua require"fzf-lua".setup({"telescope",winopts={preview={default="bat"}}})` it is working well

**Comment by ibhagwan** [comment]
> is there a 100% way of knowing which profile is being used? because I not sure about :lua require"fzf-lua".setup("fzf-native") version (version with string), it seems it still don't work properly

Not ATM, it's a new feature and it's a bunch of options sent to `setup` and since we now also added customizations on top of baseline it makese it kind of meaningless to have a "current profile".

As for `fzf-native` it just changes your previewers to `bat`:
![image](https://user-images.githubusercontent.com/59988195/219964358-d0c0b4a0-47bb-4d51-957d-f931c7938e5a.png)


vs the default builtin previewer uses a neovim window for previews:
![image](https://user-images.githubusercontent.com/59988195/219964332-d218830f-95b4-46c9-8515-be571d37961a.png)

**Comment by lumpsoid** [comment]
> Not ATM

ok, got it 


yea, I don't see bat previewer with `:lua require"fzf-lua".setup("fzf-native")`

**Comment by ibhagwan** [comment]
Didn't we just change it to: `:lua require"fzf-lua".setup({"fzf-native"})` (with the curly braces)? :-)

**Comment by lumpsoid** [comment]
...<silence>...

yea, everything is working fine

**Comment by ibhagwan** [comment]
Great, closing!

**Comment by lumpsoid** [comment]
Thank you for your help and sorry for misunderstandings

**Comment by lumpsoid** [comment]
@ibhagwan if you don't mind, can I ask you about options of fzf previewing ?

**Comment by ibhagwan** [comment]
> @ibhagwan if you don't mind, can I ask you about options of fzf previewing ?

Sure

**Comment by lumpsoid** [comment]
thanks

so, fzf has preview option `--preview-window`
and I want to fix first line, so when I preview any file, I always see first line 
```
--preview-window '~1'
```
something like this

and this option also changes preview window line, like, logic
for example
```
# without parameter
6
7
8
9
etc.

# with parameter
1
7
8
9
etc.
```
and there is some problem with interpolation, that some line would never be avalable in preview window
like with full screen scroll

and in [man page]((https://www.mankier.com/1/fzf#--preview-window) there is exmaple with offset 
```
--preview-window '~3,+{2}+3/2'
```

but on `fzf.vim` I have not been able to get this option to work properly
do you have any experience in this?

**Comment by ibhagwan** [comment]
FzfLua uses this option to set the correct preview lines when using grep/live_grep, you can always overwrite it by using the `fzf_opts` table or if you wish to send raw extra args directly to fzf you can use:
`:lua require'fzf-lua'.files({ fzf_cli_args = '…' })`

**Comment by lumpsoid** [comment]
is it supposed to be like this?
```lua
function M.insertHeadId()
    require('fzf-lua').grep({
      search = '',
      --fzf_cli_args = '--preview-window ~1',
      fzf_opts = {
          ['--preview-window'] =  '~1',
          ['--preview'] = 'bat {}'  -- parameter for understending am I doing this right
      },
...
```

**Comment by lumpsoid** [comment]
as I understand [default](https://github.com/ibhagwan/fzf-lua#default-options), yes

but I don't understend why `--preview 'bat {}'` didn't change preview window, and in profile 'fzf-native' you are using `winopts = { preview = { default = "bat" } }` and this is working

**Comment by ibhagwan** [comment]
> as I understand [default](https://github.com/ibhagwan/fzf-lua#default-options), yes
> 
> but I don't understend why `--preview 'bat {}'` didn't change preview window, and in profile 'fzf-native' you are using `winopts = { preview = { default = "bat" } }` and this is working

Few things, when using `fzf_opts` you’re sending raw args to fzf if the value has spaces you need to shell escape it, either use `= [['bat {}']]` or `= vim.fn.shellescape('bat {}')`.

as for the default preview, it’s not the same, it’s as equal as using `files.({ previewer = 'bat' })`, previewer is a an object derived from the previewers collection, it sends a callback to neovim which parses the selected entry and adjusts accordingly, for example `previewer = 'bat'` can work with `tags` which is much harder to achieve when using your simple `bat {}` preview.

**Comment by ibhagwan** [comment]
Btw, `--preview-window '~1'` - this sets the preview header line, this does not make the line offset start from 1, I was able to achieve what you described (preview always starting from 1) with:
```lua
:lua require("fzf-lua").live_grep({fzf_cli_args="--preview-window=+1"}) 
```

**Comment by lumpsoid** [comment]
yea, also just minute ago finaly get 50% what I want
`fzf_cli_args = '--preview-window=~1'`

**Comment by lumpsoid** [comment]
> `--preview-window '~1'`

oh, I see, `+1` scrolls me back to the top

**Comment by lumpsoid** [comment]
wait, `~1` locks first line in the file in preview window

**Comment by ibhagwan** [comment]
> wait, `~1` locks first line in the file in preview window

`~1` sets a “header line” that stays on top even when you scroll up/down the preview, `+1` sets the scroll to always start from line 1, not sure what your exact goal here but if it can be achieved with fzf it can be done with fzf-lua.

**Comment by lumpsoid** [comment]
thank you for help!
